actionmailer/lib/action_mailer/test_helper.rb

```diff
@@ -161,28 +162,17 @@ def assert_enqueued_email_with(mailer, method, params: nil, args: nil, queue: Ac
         mailer = mailer.instance_variable_get(:@mailer)
       end
 
-      expected_args = if args.is_a?(Proc) || params.is_a?(Proc)
-        -> (job_args) {
-          # Ordered from the least to the most heavy checks.
-          return false if [mailer.to_s, method.to_s, "deliver_now"] != job_args[0..2]
-          return false if !params.is_a?(Proc) && params != job_args[3][:params]
-          return false if !args.is_a?(Proc) && Array(args) != job_args[3][:args]
-          return false if args.is_a?(Proc) && !args.(job_args[3][:args])
-          return false if params.is_a?(Proc) && !params.(job_args[3][:params])
+      params, args = args, nil if args.is_a?(Hash)
+      args = Array(args) unless args.is_a?(Proc)
```

Comment:
Wouldn't we want to support a situation when args and params are both Hash? The original implementation didn't seem to support it, so my PR was proposing to fix it (alongside adding support for procs) using the following approach:

If either args or params are proc (something that wasn't possible before), then switch to the "new mode" of testing. In the "new mode" if you put hash in either of them, it gets tested.

Your simplification makes it such that if you try passing `args: { foo: 'bar' }, params: ->(_){ … }`, then expected _params_ are overwritten with expected _args_ and get compared against actual _params_, so the original expected params (the proc) is dropped. Additionally, expected _args_ become `[]`, so actual args are still being compared, but against an empty array. Is that desirable? Or am I misreading this? 🤔

Reply:
> Wouldn't we want to support a situation when both expected args and params are passed in, even if args is a Hash?
>
> The original implementation didn't seem to support it, so my PR was proposing to carefully fix it (I thought may've been a bug) alongside adding support for procs

I have not dug into why the original code was written that way, but it feels like a separate concern, so I think it would be better to address it in another PR.

> The bug fix would only kick in if either args or params is a proc (something that wasn't supported before). It would switch to the "new mode" of testing.

That seems unintuitive to me.  I think if we change the behavior, we should change it for all types of arguments.

Part of the reason I factored `params = ...` and `args = ...` out from the comparison logic is to make that kind of change easier, as well as implementing the suggestion from https://github.com/rails/rails/pull/46626#issuecomment-1542301076.  (But, again, as a separate PR with a dedicated discussion.)

Reply:
> I have not dug into why the original code was written that way, but it feels like a separate concern, so I think it would be better to address it in another PR. […] I think if we change the behavior, we should change it for all types of arguments.

I guess depends how you think about it.

Option 1: Add new feature "correctly", then fix old feature as a breaking change.
Option 2: Add new feature "brokenly", but consistently with old feature, then fix both of them with a breaking change for both.

I honestly prefer option 1, even though it feels a bit unintuitive. It minimizes unintended behavior.

And since this doesn't seem urgent, we also have these:

Option 3: dig into old feature, fix it, then add this new feature
Option 4: fix them all at once

Since I'm not part of the core team, I can only provide reasoning/feedback, but ultimately respect whatever decision you decide to make. Let me know if you want to go option 2, and I can help with the rebase/squash.

Reply:
The `:params` kwarg was introduced in #45752.  The lesser precedence of `:params` vs `:args` seems to be intentional.  Code that specifies `:params` is expected to specify `:args` as an array (if at all).

I think that is a reasonable expectation, so I would not call it "broken".  Though I agree it also isn't ideal.  #45752 mentions possibly deprecating the old behavior.  I have submitted #48194 to do so.

Reply:
Gotcha, yeah I was calling it "broken" for brevity, definitely a nuanced case. Nice, very efficient work. 🎉👍🏻